draft-irtf-cfrg-bls-signature.md

```diff
@@ -771,17 +771,24 @@ Outputs:
 Precondition: n >= 1, otherwise return INVALID.
 
 Procedure:
-1.  R = signature_to_point(signature)
-2.  If R is INVALID, return INVALID
-3.  If signature_subgroup_check(R) is INVALID, return INVALID
-4.  C1 = 1 (the identity element in GT)
-5.  for i in 1, ..., n:
-6.      If KeyValidate(PK_i) is INVALID, return INVALID
-7.      xP = pubkey_to_point(PK_i)
-8.      Q = hash_to_point(message_i)
-9.      C1 = C1 * pairing(Q, xP)
-10. C2 = pairing(R, P)
-11. If C1 == C2, return VALID, else return INVALID
+1  compute the l distinct messages m_1, m_l
+2. Aggregate the public keys of the same message to l sets of public keys QK_1_1, ...,QK_1_m, QK_2_1,..., QK_2_p, ..., QK_l_1,...,QK_l_q   
+3.  R = signature_to_point(signature)
+4.  If R is INVALID, return INVALID
+5.  If signature_subgroup_check(R) is INVALID, return INVALID
+6.  C1 = 1 (the identity element in GT)
+7.  for i in 1, ..., l:
+8.      aggregate = pubkey_to_point(QK_i_1)
+        for j in 2,...,len(QK_i):
+9.          next = pubkey_to_point(PK_i_j)
+10.         aggregate = aggregate + next
+11.      RK_i = point_to_pubkey(aggregate)
+12.      If KeyValidate(RK_i) is INVALID, return INVALID
+13.      xP = pubkey_to_point(RK_i)
+14.      Q = hash_to_point(m_i)
+15.      C1 = C1 * pairing(Q, xP)
+16. C2 = pairing(R, P)
+17. If C1 == C2, return VALID, else return INVALID
```

Comment:
To make sure that our fix only adds _extra_ security check, we should keep all the existing checks. The original CoreAggregateVerify calls KeyValidate for every individual public key and so we should do the same. We will add the extra check for the aggregate public key.
P/S: If we don't check individual key then there is a potential attack where individual public keys are not in the subgroup but their sum is in the subgroup.
```suggestion
8.       if KeyValidate(QK_i_1) is INVALID, return INVALID
9.       aggregate = pubkey_to_point(QK_i_1)
10.      for j in 2,...,len(QK_i):
11.         If KeyValidate(QK_i_j) is INVALID, return INVALID
12.         next = pubkey_to_point(QK_i_j)
13.         aggregate = aggregate + next
13.      RK_i = point_to_pubkey(aggregate)
14.      If len(QK_i) > 1: 
15.         If KeyValidate(RK_i) is INVALID, return INVALID
16.      xP = pubkey_to_point(RK_i)
17.      Q = hash_to_point(m_i)
18.      C1 = C1 * pairing(Q, xP)
19.  C2 = pairing(R, P)
20.  If C1 == C2, return VALID, else return INVALID
```

Reply:
@cryptosubtlety thanks a lot for your comment. I added the two validations as you suggested (it makes a lot of sense). I did not understand the need of ` If len(QK_i) > 1: ` though. May you clarify?

Reply:
thanks a lot @cryptosubtlety I have applied your suggestions.